test/test_cpp_extensions.py

```diff
@@ -106,6 +106,111 @@ def test_optional(self):
         has_value = cpp_extension.function_taking_optional(None)
         self.assertFalse(has_value)
 
+    def test_inline_jit_compile_extension_with_functions_as_list(self):
+        cpp_source = '''
+        at::Tensor tanh_add(at::Tensor x, at::Tensor y) {
+          return x.tanh() + y.tanh();
+        }
+        '''
+
+        module = torch.utils.cpp_extension.load_inline(
+            name='inline_jit_extension_with_functions_list',
+            cpp_sources=cpp_source,
+            functions='tanh_add',
+            verbose=True)
+
+        self.assertEqual(module.tanh_add.__doc__.split('\n')[2], 'tanh_add')
+
+        x = torch.randn(4, 4)
+        y = torch.randn(4, 4)
+
+        z = module.tanh_add(x, y)
+        self.assertEqual(z, x.tanh() + y.tanh())
+
+    def test_inline_jit_compile_extension_with_functions_as_dict(self):
+        cpp_source = '''
+        at::Tensor tanh_add(at::Tensor x, at::Tensor y) {
+          return x.tanh() + y.tanh();
+        }
+        '''
+
+        module = torch.utils.cpp_extension.load_inline(
+            name='inline_jit_extension_with_functions_dict',
+            cpp_sources=cpp_source,
+            functions={'tanh_add': 'Tanh and then sum :D'},
+            verbose=True)
+
+        self.assertEqual(
+            module.tanh_add.__doc__.split('\n')[2], 'Tanh and then sum :D')
+
+    def test_inline_jit_compile_extension_multiple_sources_and_no_functions(self):
+        cpp_source1 = '''
+        at::Tensor sin_add(at::Tensor x, at::Tensor y) {
+          return x.sin() + y.sin();
+        }
+        '''
+
+        cpp_source2 = '''
+        #include <torch/torch.h>
+        at::Tensor sin_add(at::Tensor x, at::Tensor y);
+        PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+          m.def("sin_add", &sin_add, "sin(x) + sin(y)");
```

Comment:
This is a bit late to ask this question (since the code for this has already been merged) but I'll ask it anyway: are we sure we want pybind11 to be part of the public interface of C++ extensions? If it is, then we are (1) always on the hook to ship pybind11 as part of PyTorch (we cannot ever decide "oh, let's use another binding library", because that will break userspace), and (2) every upgrade of pybind11 is a potentially user-visible BC-breaking change, because any BC-breakage in pybind11 is BC breakage for extension code. (Similarly, it looks like we are installing pybind11 headers as part of PyTorch, which is not very good OSS citizenship, because a downstream project might want to be using their own version of pybind11, and the headers will clobber each other.)

If we don't have concrete use-cases for letting people directly write pybind11 in the extension code, then it seems safer to (1) move our installed pybind11 headers under a private prefix (so they don't clash with a user) and (2) don't expose the functionality to directly write bindings. It's not a big loss; after all, the user can use whatever they want to get bindings, if they really want to.

Reply:
Zach and I consciously decided to expose pybind11 instead of some kind of custom `TORCH_EXTENSION` so that users could answer all their questions about bindings by looking at pybind11 docs instead of having to ask us. This makes it clear that we are not doing anything magic here, it's just pybind11. 

Your concerns are very valid but I feel like they would affect 1 out of 1,000,000 users and make the experience less pleasant for the other 999,999? For example, the extension tests, which we point to as examples of using the extension API, also have examples of binding in structs. This is even more advanced usage of pybind11, which we would have to wrap somehow, or disallow.